### PR TITLE
New block lifecycle hook and BlockToolAPI: "moved"

### LINKED ISSUE #1005: Block Tune MoveUp and MoveDown (Swap) trigger event or lifecycle hook
## Current Situation

Currently, I am using editorjs in a [react environment](https://www.npmjs.com/package/@natterstefan/react-editor-js). That is why I came up with the two other issues and questions posted earlier.

_Problem:_ It looks like there's no event I can subscribe to when a block is moved either up or down (or swapped). This leaves me clueless when a block moved.

Maybe there's an API for that already, but I was not able to find it (neither [editorjs.io/base-concepts](https://editorjs.io/base-concepts), [master/docs](https://github.com/codex-team/editor.js/tree/master/docs) or [editorjs.io/making-a-block-settings](https://editorjs.io/making-a-block-settings)).

This is critical for me (and maybe others as well), as some parts of my react component break when the block is moved. For instance, tinymce renders it's content inside a selected `div`. When the block is moved its element (iframe) is reloaded and gone as a result.*

At this point, I do not now _when_ to re-init the 3rd party div as an example.

## Expected Solution

EditorJS allows subscribing to block tune's (eg. move ups, downs and swaps) by either providing an event or a lifecycle hook.

## References and Ideas

Currently, my time's limited to contribute. But I've checked out some parts of the code and came up with the following suggestion. Adding a new lifecycle (eg. `swapped`) could be used to do whatever is needed when a block was swapped.

```diff
  // https://github.com/codex-team/editor.js/blob/ac93017c706f091fd3014c3edb248db4e03f55dd/src/components/blocks.ts#L126-L144
  public swap(first: number, second: number): void {
	// ... existing code

+    /**
+     * trigger cycle hooks
+     */
+    this.blocks[first].call(BlockToolAPI.SWAPPED); // or ON_SWAPPED
+    this.blocks[second].call(BlockToolAPI.SWAPPED);
  }
```

### Links

- [src/components/block-tunes/block-tune-move-up.ts#L59-L106](https://github.com/codex-team/editor.js/blob/ac93017c706f091fd3014c3edb248db4e03f55dd/src/components/block-tunes/block-tune-move-up.ts#L59-L106)
- [src/components/block-tunes/block-tune-move-down.ts#L58-L98](https://github.com/codex-team/editor.js/blob/ac93017c706f091fd3014c3edb248db4e03f55dd/src/components/block-tunes/block-tune-move-down.ts#L58-L98)
- [src/components/block.ts#L31-L44](https://github.com/codex-team/editor.js/blob/ac93017c706f091fd3014c3edb248db4e03f55dd/src/components/block.ts#L31-L44)
- [src/components/blocks.ts#L126-L144](https://github.com/codex-team/editor.js/blob/ac93017c706f091fd3014c3edb248db4e03f55dd/src/components/blocks.ts#L126-L144)
- [src/components/block.ts#L418-L437](https://github.com/codex-team/editor.js/blob/ac93017c706f091fd3014c3edb248db4e03f55dd/src/components/block.ts#L418-L437)
- [src/components/dom.ts#L129-L149](https://github.com/codex-team/editor.js/blob/ac93017c706f091fd3014c3edb248db4e03f55dd/src/components/dom.ts#L129-L149)
- [editorjs.io/listeners](https://editorjs.io/listeners)

\* Context: the 3rd party tool renders its content inside an `iframe`, and according to [this post on stackoverflow](https://stackoverflow.com/a/8318401/1238150) `iframe`s update when they were moved in the dom. This was an issue mentioned in the [tinymce repo years ago](https://github.com/tinymce/tinymce/issues/1073).

### PR DESCRIPTION
**Pull request by natterstefan**
## Related Issue

resolves #1005, which describes an example use-case for this feature in more details.

## How to test

```sh
cd example/tools/header
code src/index.js
```

add the following code at the bottom (near line 464)

```js
  moved(event) {
    console.log('>> I WAS MOVED', event);
  }
```

and run `yarn build`.

Finally, open the example page with `open example/example-dev.html` and swap the header tool.

## Notes

This is my very first contribution to https://github.com/codex-team/editor.js, and I need to get used to your plans about lifecycles and callbacks. So please, be gentle ;)

~~Additionally, it might be helpful when there's a "How to develop" section in the README. After, checking the `package.json`'s `scripts` section I had an idea how to start though.~~ ==> https://editorjs.io/core-development

## Example

Screen Capture on 2020-01-10 at 20-52-21

https://d.pr/i/SUOMHq/9TjBiEy2mF

### CODE REVIEW

**Comment by natterstefan** [comment]
@neSpecc @khaydarov @gohabereg Could you please review this feature? Our customers and apps definitely need this one. Thanks for your help and time! Highly appreciated.

**Comment by gohabereg** [comment]
Hi @natterstefan 
PR is OK. However, we discussed with the team and decided that `swap` operation is logically incorrect. Hook should be triggered only for block user moves. So operation should be `move(block, index)` (as well as hook should be `MOVED`) and should use `insertAdjacentElement` and `splice` methods for DOM and `blocks` array respectively.

Could you implement these changes yourself?

**Comment by natterstefan** [comment]
Hi @gohabereg, as discussed I deprecated `swap` and added `move` instead. Let me know what you think.

**Comment by natterstefan** [comment]
FTR: I am entirely happy with the wording `MoveEvent` but `MoveResult` seemed wrong as well ==> https://github.com/codex-team/editor.js/pull/1007/files#diff-c1c1acd51e3a8ab70e7639423cb0f08fR480. Maybe you can come up with a better wording for that.

File: src/components/blocks.ts [hunk 1]
```diff
@@ -143,6 +145,32 @@ export default class Blocks {
     this.blocks[first] = secondBlock;
   }
 
+  /**
+   * Move a block from one to another index
+   * @param {Number} toIndex - new index of the block
+   * @param {Number} fromIndex - block to move
+   */
+  public move(toIndex: number, fromIndex: number): void {
+    const block = this.blocks.splice(fromIndex, 1)[0];
```
Comment (gohabereg) [inline src/components/blocks.ts]: Is there any reason to use `splice` instead of `this.blocks[fromIndex]`?

File: src/components/blocks.ts [hunk 2]
```diff
@@ -143,6 +145,32 @@ export default class Blocks {
...
+
+    // manipulate DOM
+    const prevIndex = toIndex - 1;
+    const previousBlockIndex = prevIndex < 0 ? 0 : prevIndex;
+    const previousBlock = this.blocks[previousBlockIndex];
+    if (toIndex > 0) {
```
Comment (gohabereg) [inline src/components/blocks.ts]:
```suggestion
    
    if (toIndex > 0) {
```

File: src/components/blocks.ts [hunk 3]
```diff
@@ -143,6 +145,32 @@ export default class Blocks {
...
+    } else {
+      this.insertToDOM(block, 'beforebegin', previousBlock);
+    }
+
+    // move in array
+    this.blocks.splice(toIndex, 0, block);
+
+    // invoke hook
+    const event: MoveEvent = { fromIndex, toIndex };
+    block.call(BlockToolAPI.MOVED, event);
```
Comment (gohabereg) [inline src/components/blocks.ts]: This is ok as well:
```
block.call(BlockAPI.MOVED, { fromIndex, toIndex });
```

File: src/components/modules/blockManager.ts [hunk 4]
```diff
@@ -561,6 +562,30 @@ export default class BlockManager extends Module {
...
 
+  /**
+   * Move a block to a new index
+   * @param {Number} toIndex
+   * @param {Number} fromIndex
+   */
+  public move(toIndex, fromIndex = this.currentBlockIndex): void {
+    // make sure indexes are valid and within a valid range
+    if (isNaN(toIndex) || isNaN(fromIndex)) {
+      _.log(`Error during 'move' call: incorrect indices provided.`, 'error');
```
Comment (gohabereg) [inline src/components/modules/blockManager.ts]: I guess `warn` is enough here

File: src/components/modules/blockManager.ts [hunk 5]
```diff
@@ -561,6 +562,30 @@ export default class BlockManager extends Module {
...
+      return;
+    }
+
+    if (!this.validateIndex(toIndex) || !this.validateIndex(fromIndex)) {
+      _.log(`Error during 'move' call: indices cannot be lower than 0 or greater than the amount of blocks.`, 'error');
```
Comment (gohabereg) [inline src/components/modules/blockManager.ts]: And here as well

File: src/components/modules/blockManager.ts [hunk 6]
```diff
@@ -604,4 +629,16 @@ export default class BlockManager extends Module {
...
+  /**
+  * validates given index is not lower than 0 or higher than the amount of
+  * blocks
+  * @param {number} index
+  */
+  private validateIndex(index: number) {
+    if (index < 0 || index >= this._blocks.length) {
+      return false;
+    }
+    return true;
```
Comment (gohabereg) [inline src/components/modules/blockManager.ts]:
```suggestion

    return true;
```

File: types/tools/tune-events.d.ts [hunk 7]
```diff
@@ -0,0 +1,4 @@
+export type MoveEvent = {
```
Comment (gohabereg) [inline types/tools/tune-events.d.ts]: This is relevant not only for tunes, so `hook-events` is better name for file I guess

**Comment by gohabereg** [comment]
> FTR: I am not entirely happy with the wording `MoveEvent` but `MoveResult` seemed wrong as well ==> https://github.com/codex-team/editor.js/pull/1007/files#diff-c1c1acd51e3a8ab70e7639423cb0f08fR480. Maybe you have a better idea and wording for that.

Well, it's not actually an event. However, I think it's ok

File: src/components/blocks.ts (hunk 3 above)
Comment (natterstefan) [reply src/components/blocks.ts]: That's right, 👍 I just wanted to have the type safety here.

File: src/components/blocks.ts (hunk 1 above)
Comment (natterstefan) [reply src/components/blocks.ts]: As discussed, I added some notes here https://github.com/codex-team/editor.js/pull/1007/commits/1551be3a2f46f97233004e1396376a2353d02ab2#diff-a1e3d44548903047b0a0d542527a389cR154-R158.

**Comment by Infonautica** [comment]
I wait for this feature so much!
Suggestions for MoveEvent naming:
Result, indexes, changes, info, details, updates

File: src/components/blocks.ts [hunk 8]
```diff
@@ -143,6 +145,38 @@ export default class Blocks {
...
```
Comment (gohabereg) [inline src/components/blocks.ts]: Add please an empty line after `const` expression

File: types/tools/hook-events.d.ts [hunk 9]
```diff
@@ -0,0 +1,4 @@
+export type MoveEvent = {
```
Comment (neSpecc) [inline types/tools/hook-events.d.ts]: please write a docs for this object and properties

File: src/components/modules/blockManager.ts [hunk 10]
```diff
@@ -604,4 +629,17 @@ export default class BlockManager extends Module {
     Listeners.on(block.holder, 'dragover', (event) => BlockEvents.dragOver(event as DragEvent));
     Listeners.on(block.holder, 'dragleave', (event) => BlockEvents.dragLeave(event as DragEvent));
   }
+
+  /**
+  * validates given index is not lower than 0 or higher than the amount of
```
Comment (neSpecc) [inline src/components/modules/blockManager.ts]:
```suggestion
  * Validates that the given index is not lower than 0 or higher than the amount of
```

File: src/components/blocks.ts [hunk 11]
```diff
@@ -143,6 +145,38 @@ export default class Blocks {
...
+    /**
+     * cut out the block, move the DOM element and insert at the desired index
+     * again (the shifting within the blocks array will happen automatically).
+     * @see https://stackoverflow.com/a/44932690/1238150
+     */
+    const block = this.blocks.splice(fromIndex, 1)[0];
+
+    // manipulate DOM
+    const prevIndex = toIndex - 1;
+    const previousBlockIndex = prevIndex < 0 ? 0 : prevIndex;
```
Comment (neSpecc) [inline src/components/blocks.ts]:
```suggestion
    const previousBlockIndex = Math.max(0, prevIndex);
```

File: src/components/modules/blockManager.ts [hunk 12]
```diff
@@ -604,4 +629,17 @@ export default class BlockManager extends Module {
...
+  * blocks
+  * @param {number} index
+  */
+  private validateIndex(index: number) {
```
Comment (neSpecc) [inline src/components/modules/blockManager.ts]:
```suggestion
  private validateIndex(index: number): boolean {
```

**Comment by natterstefan** [comment]
Hello @gohabereg, hello @neSpecc.

I've updated the PR, according to your review _and_ changed the `MoveEvent`. It represents now more the shape of an event. See https://github.com/codex-team/editor.js/pull/1007/commits/ae07a65f2cc79ea7134dd7d2de0c082799aa8cc8#diff-0e4c6bcb0a2e8a6a478c9c39ba9531fe, please.

image

File: types/tools/hook-events.d.ts [hunk 13]
```diff
@@ -0,0 +1,20 @@
...
+   * index the block was moved to
+   */
+  toIndex: number;
+}
+
+/**
+ * Move event for block relocation
+ */
+export interface MoveEvent {
+  readonly detail: MoveEventDetail;
```
Comment (neSpecc) [inline types/tools/hook-events.d.ts]: I don't quite understand, why we need this `detail` wrapper?

Comment (gohabereg) [reply types/tools/hook-events.d.ts]: This is common `Event` structure

Comment (Infonautica) [reply types/tools/hook-events.d.ts]: Not simple Event i suppose, but CustomEvent structure:
https://developer.mozilla.org/en-US/docs/Web/API/CustomEvent/CustomEvent

Comment (neSpecc) [reply types/tools/hook-events.d.ts]: @Infonautica yes, that is what I'm talking about. 
@natterstefan if you want to create an event-like structure, it's better to use CustomEvent.

**Comment by neSpecc** [comment]
@natterstefan hi, what about this request? It requires little improvements.

**Comment by natterstefan** [comment]
Hi, I'm sorry Peter. It's definitely on my todo list and I know what you
mean and what’s left. :) ok?
-- 

Mit freundlichen Grüßen,
Best regards,
Stefan

--------
*STEFAN NATTER*
Software Engineer @ Lovely Systems
<http://lovelysystems.com/?utm_source=natterstefan&utm_medium=signatur>
Website <https://natterstefan.me/> I Twitter
<https://twitter.com/natterstefan> I Linkedin
<http://www.linkedin.com/in/natterstefan> I stefan@natter.at

**Comment by neSpecc** [comment]
@natterstefan sure, thanks.

**Comment by khaydarov** [comment]
@natterstefan thank you, merged it. Checkout into release2.17

**Comment by natterstefan** [comment]
> @natterstefan thank you, merged it. Checkout release2.17

thanks for completing the feature and branch! Awesome, can't wait to use this new feature. 🤘🏻

**Pull request merged**